Feature/iglistkit implementation

polikhill opened the pull request.

polikhill committed:
setup iglistkit; bug with cell;

polikhill committed:
fix outlets;

polikhill committed:
fix cell selection; fix refreshControl;

aLHF commented on ReduxTestTask/Base/BaseClasses/BaseCollectionViewCell.swift:
```diff
@@ -0,0 +1,21 @@
+//  BaseCollectionViewCell.swift
+//  ZablotskyDental
+//
+//  Created by Dima Ripa on 8/27/18.
```
🙂

aLHF commented on ReduxTestTask/Classes/NewsList/View/NewsParentController.swift:
```diff
@@ -0,0 +1,133 @@
+    
+    private func render(props: Props) {
+        if renderedProps?.items != props.items {
+            data = props.items.compactMap({ DiffableBox(value: $0, identifier: $0.diffIdentifier as NSObjectProtocol, equal: ==) })
```
DiffableBox(...) should be created in makeProps function.

aLHF commented on ReduxTestTask/Classes/NewsList/View/NewsParentController.swift:
```diff
@@ -0,0 +1,133 @@
+    }
+    
+    private lazy var contentView = NewsListView()
+    private var data = [ListDiffable]()
```
This data should be in the view. 

Controller and view have different props. Like this:

Props {
  let viewProps: NewsListView.Props
  let error: String?
  let isLoading: Bool
}

aLHF commented on ReduxTestTask/Classes/NewsList/ViewModel/Middleware/NewsListFetchMiddleware.swift:
```diff
@@ -93,6 +93,7 @@ extension NewsList {
             Observable
                 .just(article)
                 .map ({ article -> Action in
+                    showArticle.onNext(article)
```
map should be used for transformations, not for side effects. Same for flatMap.

Use do(onNext:) for side effects

aLHF commented on ReduxTestTask/Classes/NewsList/ViewModel/NewsListStore.swift:
```diff
@@ -67,8 +66,8 @@ extension NewsList {
             
-        case let action as ShowArticle:
-            newState.selectedArticle = action.article
+        case is ShowArticle:
```
let's remove this if we don't do any work here. We have the default case for this. Same for isSelectedCell

aLHF commented on ReduxTestTask/Classes/NewsList/ViewModel/NewsListViewModel.swift:
```diff
@@ -55,16 +56,29 @@ extension NewsList {
                 .voidValues()
+    
+        
+//            let article = store.state
```
let's remove commented code

polikhill committed:
remove empty cases in NewsListStore;
remove commented code in NewsListViewModel;
add do(onNext: ) in  makeShowArticleMiddleware(showArticle:)

polikhill committed:
Merge branch 'feature/newsList' into feature/iglistkit_implementation

# Conflicts:
#	ReduxTestTask.xcodeproj/project.pbxproj
#	ReduxTestTask/App/AppDelegate.swift
#	ReduxTestTask/Classes/NewsList/View/NewsListView.swift
#	ReduxTestTask/Classes/NewsList/View/NewsListViewController.swift
#	ReduxTestTask/Classes/NewsList/ViewModel/NewsListProps.swift
#	ReduxTestTask/Classes/NewsList/ViewModel/NewsListViewModel.swift

polikhill committed:
fix render(props: ) in NewsListViewController;

polikhill committed:
replace adapter to NewsListView;
refactor makeShowArticleMiddleware(showArticle: );

aLHF commented on ReduxTestTask/Classes/NewsList/View/NewsListView.swift:
```diff
@@ -9,50 +9,80 @@
     private func toggleLoading(on: Bool) {
         if on {
-            refreshControl.beginRefreshing()
+            DispatchQueue.main.async {
```
render should be called on UI Thread so we don't need to dispatch here

aLHF commented on ReduxTestTask/Classes/NewsList/ViewModel/NewsListProps.swift:
```diff
@@ -14,17 +14,18 @@ extension NewsList {
             contentViewProps: NewsListView.ContentViewProps(
                 isLoading: state.isLoading,
-                items: state.fetchedNews.compactMap { makeNewsCellProps(from: $0) }),
+                items: state.fetchedNews.compactMap { makeNewsCellDiffableProps(from: $0) }),
```
why compactMap?

polikhill replied:
I forgot that compactMap for optional init

polikhill committed:
remove dispatch in toggleLoading(on: );

aLHF commented on ReduxTestTask/Classes/NewsList/View/NewsListController.swift:
```diff
@@ -0,0 +1,60 @@
+        ) -> UICollectionViewCell & ListBindable {
+
+        
+        guard let cell = collectionContext?.dequeueReusableCell(withNibName: "\(NewsCell.self)", bundle: nil, for: self, at: index) as? NewsCell else { fatalError() }
```
you have an extension for tableView or collectionView with a helper for dequeue. Let's write one for IGListKit too

polikhill committed:
add IGListKitExtension for collectionContext;
